SSO post

sebadoom opened the pull request:
Draft, missing example and diagrams. Working on the example.

sebadoom commented:
WiP
![understanding single-sign-on for the web](https://cloud.githubusercontent.com/assets/773367/9948726/ff0c58cc-5d7b-11e5-8fd4-993e6d14e0e0.png)
![Uploading Understanding Single-Sign-On for the Web.png…]()

mgonto commented on _posts/2015-09-18-understanding-single-sign-on-for-the-web.markdown:
```diff
@@ -0,0 +1,84 @@
+- credentials
+---
+
+Nowadays, almost every website requires some form of authentication to access its features and content. With the number of websites and services only rising, a centralized login system has become a necessity. Single-sign-on is now required more than ever. In this post we will study how single-sign-on is implemented for the web. We will also compare it to mobile and intranet implementations so you ...
```
I'd explain in here in 4 lines what Single Sign On is. And then do a deeper dive below. This intro doesn't tell me what I'll learn about in the post, other than something called Single Sign On. WDYT?

mgonto commented on _posts/2015-09-18-understanding-single-sign-on-for-the-web.markdown:
```diff
@@ -0,0 +1,84 @@
+- User attributes exchange
+- User management
+
+The **authentication** aspect deals with validating user credentials and establishing the identity of the user. **Authorization** is related to access restrictions (is the user allowed to access X resource?). The **attributes exchange** part deals with data sharing across different user management systems. For instance, fields such as *"real name"* may be present in both systems. A federated iden ...
```
typo in name"*

mgonto commented on _posts/2015-09-18-understanding-single-sign-on-for-the-web.markdown:
```diff
@@ -0,0 +1,84 @@
+- External or internet SSO
+- Mobile SSO
+
+The main difference between the first two lies in how easy it is to **modify a preexisting system** to share authentication details across other systems. Internal or intranet systems can usually be modified to communicate using one of the many different SSO implementations. For example, suppose two big intranet sites were developed by different teams: one uses a simple custom user + password auth ...
```
This text isn't really clear. The main difference is that if it's internal SSO you'd probably just hook it to an existing app using some protocol as you say. Do the example with SAML and mention 2 real services like Dropbox and Slack for example. SAML is more common than OAuth 2 for this.

mgonto commented on _posts/2015-09-18-understanding-single-sign-on-for-the-web.markdown:
```diff
@@ -0,0 +1,84 @@
+
+The main difference between the first two lies in how easy it is to **modify a preexisting system** to share authentication details across other systems. Internal or intranet systems can usually be modified to communicate using one of the many different SSO implementations. For example, suppose two big intranet sites were developed by different teams: one uses a simple custom user + password auth ...
+
+For external or internet systems things are a bit more difficult: there is usually not much flexibility. If no modifications are possible, SSO is usually implemented by **converting data** between authentication solutions. For instance, OAuth2 data may be converted to SAML by a service.
```
For External or Internet ones, we should mention that they're usually custom built and implementing it in custom built ones is harder to do manually

mgonto commented on _posts/2015-09-18-understanding-single-sign-on-for-the-web.markdown:
```diff
@@ -0,0 +1,84 @@
+- Use one of the SSO solutions provided by the OS (or a third party). For instance, since iOS 7, Apple provides a feature called *Enterprise Single-Sign-On*.
+
+### Client-side vs server-side SSO for the web
+A typical web application stores sessions using **cookies** (or other forms of **client-side storage**). Web browsers provide mechanisms to ensure session information cannot be stolen by other domains. Therefore, SSO needs to deal with this in some way to do its main functionality. How it deals with that usually depends on the type of web application being developed. There are usually two types o ...
```
Explain this by mentioning first an example. let's imagine company A which has a.com buys another company and now has b.com. They want users to be shared..... Let's tell a story :)

mgonto replied:
Actually use the google.com example that's better

mgonto commented on _posts/2015-09-18-understanding-single-sign-on-for-the-web.markdown:
```diff
@@ -0,0 +1,84 @@
+### Client-side vs server-side SSO for the web
+A typical web application stores sessions using **cookies** (or other forms of **client-side storage**). Web browsers provide mechanisms to ensure session information cannot be stolen by other domains. Therefore, SSO needs to deal with this in some way to do its main functionality. How it deals with that usually depends on the type of web application being developed. There are usually two types o ...
+
+Regular (many-page) web apps do SSO by storing session data in a cookie from a **single domain**. In other words, whenever a user needs to login from a domain that is part of the SSO system, he or she is sent to the domain that performs the authentication. Information about the **requested resource** from the original domain is usually attached to this request. After the user validates his identi ...
```
Continue with auth0.com example. All the cookies for both are saved in auth0.com

mgonto commented on _posts/2015-09-18-understanding-single-sign-on-for-the-web.markdown:
```diff
@@ -0,0 +1,84 @@
+
+TODO: SERVER SIDE SSO DIAGRAM
+
+In contrast, single page apps usually handle authentication through **AJAX** calls. Since AJAX is limited to the *same domain policy* by web browsers, **cross-origin resource sharing (CORS)** is necessary. With CORS, direct requests to an authentication server can be performed without reloads or redirects. Authentication information may then be requested for each application and saved as a cookie ...
```
This is not clear. When it's client side SSO they can still do the redirect as in Regular Web App if they see that the central server has the cookie.

Basically you do a CORS request to auth0.com, auth0.com reads the cookie and returns that in the body. If there's something in there, then it can do SSO, which means that we need to redirect to AUth0.com, it'll automatically authenticate us and redirect back to the SPA with the JWT in the header so it's different that what's stated here. @woloski can you confirm what I'm saying is OK

mgonto commented on _posts/2015-09-18-understanding-single-sign-on-for-the-web.markdown:
```diff
@@ -0,0 +1,84 @@
+TODO: CLIENT SIDE SSO DIAGRAM
+
+### Different frameworks
+If you have been reading about SSO online, you will probably find there are many different implementations: OpenID Connect, Facebook Connect, SAML, Microsoft Account (formerly known as Passport), etc. Our advice is: choose whatever is simpler for your development efforts. For instance, SAML is deeply entrenched in enterprise developments, so in some cases it will make sense to pick that. If you t ...
```
They're not frameworks, they're protocols

mgonto commented on _posts/2015-09-18-understanding-single-sign-on-for-the-web.markdown:
```diff
@@ -0,0 +1,84 @@
+TODO
+
+## Aside: SSO with Auth0
+If you are already using Auth0 in your developments, you know how easy it is to do SSO. If not, see the [docs](https://auth0.com/docs/sso/single-sign-on) and check the [examples](https://github.com/auth0/auth0-sso-sample). Our SSO solution works as a *bridge* between different SSO frameworks. So, whatever your existing apps are using, it has never been easier to integrate SSO in them.
```
We've explained how it works if you do the changes upsteairs so you can mention that in here :D

sebadoom commented:
Final draft, implemented comments by @woloski, removed example (converted into post series, mentions example as part of the next post). Added diagrams (awaiting for approval before requesting versions from evilrabbit).
![understanding single-sign-on for the web part 1 1](https://cloud.githubusercontent.com/assets/773367/10048747/c9b8845c-61ea-11e5-9de5-93b8dd65f9d1.png)

sebadoom commented:
New diagrams, wordy changes.
![understanding single-sign-on for the web part 1](https://cloud.githubusercontent.com/assets/773367/10053637/9eeafd04-6203-11e5-86fb-7f2d643d5f40.png)

sebadoom commented:
Final version.
![what is and how does single sign on work](https://cloud.githubusercontent.com/assets/773367/10056532/181189d8-6212-11e5-9287-c0de86d72da1.png)
